File: backend/app/schemas/mft.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class MftFileOut(BaseModel):
    id:          str
    case_id:     str
    filename:    str
    status:      str
    entry_count: int | None
    error_msg:   str | None
    uploaded_at: datetime
    parsed_at:   datetime | None
    added_to_evidence:     bool
    parse_progress:        int
    parse_duration_seconds: int | None

    model_config = {"from_attributes": True}


class MftEntryOut(BaseModel):
    """One MFT record as returned by the API (sourced from DuckDB / MFTECmd CSV)."""
    entry_number:        int
    parent_entry_number: int | None
    parent_path:         str | None
    filename:            str | None
    extension:           str | None
    file_size:           int | None

    is_in_use:   bool
    is_deleted:  bool
    is_directory: bool

    si_created:     datetime | None
    si_modified:    datetime | None
    si_accessed:    datetime | None
    si_mft_changed: datetime | None

    fn_created:     datetime | None
    fn_modified:    datetime | None
    fn_accessed:    datetime | None
    fn_mft_changed: datetime | None

    has_ts_anomaly: bool   # SI timestamps < FN timestamps (possible timestomping)


class EntriesPage(BaseModel):
    total:     int
    page:      int
    page_size: int
    pages:     int
    items:     list[MftEntryOut]


class MftSummary(BaseModel):
    total_entries:      int
    deleted_count:      int
    directory_count:    int
    file_count:         int
    oldest_si_modified: datetime | None
    newest_si_modified: datetime | None
    top_extensions:     list[dict]   # [{ext, count}]

File: frontend/src/api/mft.ts
import api from './client'

export interface MftFile {
  id:                     string
  case_id:                string
  filename:               string
  status:                 'pending' | 'parsing' | 'ready' | 'error'
  entry_count:            number | null
  error_msg:              string | null
  uploaded_at:            string
  parsed_at:              string | null
  added_to_evidence:      boolean
  parse_progress:         number
  parse_duration_seconds: number | null
}

export interface MftEntry {
  entry_number:        number
  parent_entry_number: number | null
  parent_path:         string | null
  filename:            string | null
  extension:           string | null
  file_size:           number | null

  is_in_use:   boolean
  is_deleted:  boolean
  is_directory: boolean

  si_created:     string | null
  si_modified:    string | null
  si_accessed:    string | null
  si_mft_changed: string | null

  fn_created:     string | null
  fn_modified:    string | null
  fn_accessed:    string | null
  fn_mft_changed: string | null

  has_ts_anomaly: boolean
}

/** MFT entry enriched with source file metadata for the selection panel. */
export interface PinnedMftEntry extends MftEntry {
  _key:        string   // `${fileId}:mft:${entry_number}`
  _fileId:     string
  _filename:   string
  _sourceType: 'mft'
}

export interface EntriesPage {
  total:     number
  page:      number
  page_size: number
  pages:     number
  items:     MftEntry[]
}

export interface MftSummary {
  total_entries:      number
  deleted_count:      number
  directory_count:    number
  file_count:         number
  oldest_si_modified: string | null
  newest_si_modified: string | null
  top_extensions:     Array<{ ext: string; count: number }>
}

export interface EntryFilters {
  page?:       number
  page_size?:  number
  search?:     string
  flags?:      string
  extension?:  string
  time_field?: string
  time_from?:  string
  time_to?:    string
  sort_by?:    string
  sort_dir?:   string
}

export const mftApi = {
  upload: (caseId: string, file: File) => {
    const fd = new FormData()
    fd.append('file', file)
    return api.post<MftFile>(`/mft/${caseId}/upload`, fd, {
      headers: { 'Content-Type': 'multipart/form-data' },
    }).then(r => r.data)
  },

  listFiles: (caseId: string) =>
    api.get<MftFile[]>(`/mft/${caseId}/files`).then(r => r.data),

  deleteFile: (caseId: string, fileId: string) =>
    api.delete(`/mft/${caseId}/files/${fileId}`),

  summary: (caseId: string, fileId: string) =>
    api.get<MftSummary>(`/mft/${caseId}/files/${fileId}/summary`).then(r => r.data),

  entries: (caseId: string, fileId: string, filters: EntryFilters = {}) =>
    api.get<EntriesPage>(`/mft/${caseId}/files/${fileId}/entries`, { params: filters })
       .then(r => r.data),

  reparse: (caseId: string, fileId: string) =>
    api.post<MftFile>(`/mft/${caseId}/files/${fileId}/reparse`).then(r => r.data),

  addEvidence: (caseId: string, fileId: string) =>
    api.post<MftFile>(`/mft/${caseId}/files/${fileId}/add-evidence`).then(r => r.data),
}
